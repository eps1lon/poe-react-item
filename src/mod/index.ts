export { default as Simple } from './Simple';
export { default as Extended } from './Extended';
