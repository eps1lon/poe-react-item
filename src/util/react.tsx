import * as React from 'react';

// text nodes dont need keys
export const CommaSeparator = () => ', ';

export const LineBreak = (key: string) => <br key={key} />;

export function intersperse(
  nodes: React.ReactNode[],
  renderSeparator: (key: string) => React.ReactNode = CommaSeparator,
) {
  if (nodes.length === 0) {
    return [];
  }

  return nodes.slice(1).reduce(
    (joined: React.ReactNode[], element, index) => {
      // push instead of concat's returnval because concat is expensive
      joined.push(renderSeparator(`separator-${index}`), element);
      return joined;
    },
    [nodes[0]],
  );
}
