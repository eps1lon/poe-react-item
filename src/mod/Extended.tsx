import { PureComponent } from 'react';

import Mod, { Props as ModProps } from './Component';

export interface Props {
  mods: ModProps[];
}

export default class Extended extends PureComponent<Props> {
  render() {
    return 'ExtendedMods';
  }
}
