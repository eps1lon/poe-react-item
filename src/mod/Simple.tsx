import { PureComponent } from 'react';

import Mod, { Props as StatProps } from '../stat/Stat';

export interface Props {
  stats: StatProps[];
}

export default class Simple extends PureComponent<Props> {
  render() {
    return 'SimpleMods';
  }
}
