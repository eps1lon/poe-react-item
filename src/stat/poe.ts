import { Value } from '../util/value';

export interface Stat {
  id: string;
  value: Value;
}
