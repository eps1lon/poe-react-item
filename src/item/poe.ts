import { ReactNode } from 'react';

import { Mod } from '../mod/poe';
import { SingleValue, Value, AugmentableValue } from '../util/value';

// use intersection here instead of extends to be able to test Properties
// with only e.g. ArmourProperties
export type Item = AbstractItem & Properties;

export interface AbstractItem {
  base: BaseItem;
  // only required for rare and unique items
  name?: string;
  rarity: Rarity;
  requirements?: Requirements;
  implicits?: Affixes;
  explicits?: Affixes;
  // in case only stats are given and the item is magic we need to know
  // the name of the prefix/suffix mod
  prefix?: string;
  suffix?: string;
  elder?: boolean;
  shaper?: boolean;
}

export type Properties =
  | ShieldProperties
  | ArmourProperties
  | WeaponProperties
  | NoProperties;

export interface AbstractProperties {
  quality?: number;
}

export interface ArmourProperties extends AbstractProperties {
  armour?: AugmentableValue<SingleValue>;
  energy_shield?: AugmentableValue<SingleValue>;
  evasion?: AugmentableValue<SingleValue>;
}

export interface ShieldProperties extends ArmourProperties {
  // in percent
  block: AugmentableValue<SingleValue>;
}

export interface WeaponProperties extends AbstractProperties {
  physical_damage?: AugmentableValue;
  cold_damage?: Value;
  fire_damage?: Value;
  lightning_damage?: Value;
  chaos_damage?: Value;
  // time between attacks in ms
  attack_time?: AugmentableValue<SingleValue>;
  /**
   * value: 0-10000
   *        10000 => 100%
   *        1000 => 10%
   *        100 => 1%
   *        10 => 0.1%
   *        1 => 1%
   */
  crit?: AugmentableValue<SingleValue>;
}

export type NoProperties = AbstractProperties;

export interface BaseItem {
  name: string;
}

export enum Rarity {
  normal,
  magic,
  rare,
  unique,
}

export interface Requirements {
  [key: string]: AugmentableValue<SingleValue> | undefined;
  level?: AugmentableValue<SingleValue>;
  dexterity?: AugmentableValue<SingleValue>;
  inteliligence?: AugmentableValue<SingleValue>;
  strength?: AugmentableValue<SingleValue>;
}

export type Affix = ReactNode | Mod;
export type Affixes = Affix[];
