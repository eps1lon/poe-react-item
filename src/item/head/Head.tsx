import classnames from 'classnames';
import * as React from 'react';

import { ItemProps, Rarity } from '../poe';

import TypeLine from './TypeLine';
import DoubleLine from './DoubleLine';

export interface Props {
  item: ItemProps;
}

export default class Head extends React.PureComponent<Props> {
  render() {
    const { item } = this.props;
    const single_line = hasSingleNameLine(item);

    return (
      <header className={classnames({ 'double-line': !single_line })}>
        {single_line ? <TypeLine item={item} /> : <DoubleLine item={item} />}
      </header>
    );
  }
}

function hasSingleNameLine(item: ItemProps) {
  return item.rarity === Rarity.magic || item.rarity === Rarity.normal;
}
