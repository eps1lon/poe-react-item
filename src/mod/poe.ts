import { Stat } from '../stat/poe';
import { Value } from '../util/value';

export enum Group {
  implicit,
  explicit,
  enchantment,
  crafted,
}

export enum Type {
  prefix,
  suffix,
  unique,
}

export const isPrefix = (mod: Mod) => mod.type === Type.prefix;
export const isSuffix = (mod: Mod) => mod.type === Type.suffix;

export interface Mod {
  id: string;
  type?: Type;
  name?: string;
  stats?: Stat[];
}
