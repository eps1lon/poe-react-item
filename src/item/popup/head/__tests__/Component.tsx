import * as React from 'react';
import renderer from 'react-test-renderer';

import Head from '../Component';
import { Type as ModType } from '../../../../mod/poe';
import { Item as ItemProps } from '../../../poe';
import { Popup } from '../../../index';
import { Item, Rarity } from '../../../poe';

it('should display normal', () => {
  expect(
    renderer
      .create(
        <Head
          item={{
            base: {
              name: 'Full Scale Armour',
            },
            rarity: Rarity.normal,
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display magic', () => {
  expect(
    renderer
      .create(
        <Head
          item={{
            base: {
              name: 'Blinder',
            },
            rarity: Rarity.magic,
            explicits: [
              {
                type: ModType.prefix,
                name: 'Buzzing',
                stats: [],
              },
              {
                type: ModType.suffix,
                name: 'of Mastery',
                stats: [],
              },
            ],
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display rare', () => {
  expect(
    renderer
      .create(
        <Head
          item={{
            base: {
              name: 'Vaal Regalia',
            },
            name: 'Gale Salvation',
            rarity: Rarity.rare,
            explicits: [
              { type: ModType.prefix, id: 'Sapphire', name: 'Sapphire' },
              {
                type: ModType.prefix,
                id: 'LocalIncreasedEnergyShieldPercentAndStunRecovery3',
                name: "Boggart's",
              },
              {
                type: ModType.prefix,
                id: 'IncreasedEnergyShieldPercent2',
                name: 'Strong-Willed',
              },
              {
                type: ModType.suffix,
                id: 'FireResist6',
                name: 'of the Volcano',
              },
            ],
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});
