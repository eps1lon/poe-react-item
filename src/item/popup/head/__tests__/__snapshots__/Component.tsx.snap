// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display magic 1`] = `
<header
  className=""
>
  <span
    className="name-left"
  />
  <div
    className="item-name type-line"
  >
    <span>
      Buzzing 
      <span>
        Blinder
      </span>
       of Mastery
    </span>
  </div>
  <span
    className="name-right"
  />
</header>
`;

exports[`should display normal 1`] = `
<header
  className=""
>
  <span
    className="name-left"
  />
  <div
    className="item-name type-line"
  >
    <span>
      Full Scale Armour
    </span>
  </div>
  <span
    className="name-right"
  />
</header>
`;

exports[`should display rare 1`] = `
<header
  className="double-line"
>
  <span
    className="name-left"
  />
  <div
    className="item-name"
  >
    Gale Salvation
  </div>
  <div
    className="item-name type-line"
  >
    <span>
      Vaal Regalia
    </span>
  </div>
  <span
    className="name-right"
  />
</header>
`;
